src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -795,55 +807,64 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
       val pis = StringUtils.leftPad(is, d, "0")
       val filename = path + "/parts/part-" + pis
 
-      val dos = new DataOutputStream(sHadoopBc.value.value.unsafeWriter(filename))      
+      val dos = new DataOutputStream(sHadoopBc.value.value.unsafeWriter(filename))
       dos.writeInt(nRowsInBlock)
       dos.writeInt(nColsInBlock)
       dos.writeBoolean(true) // transposed, stored row major
-      
+
       dos
     }
-    
-    assert((gp.nCols << 3) <= Int.MaxValue)
-    val bytes = new Array[Byte]((gp.nCols << 3).toInt)
-    
+
+    val bytes = new Array[Byte](blockSize << 3)
+
+    val ur = new UnsafeRow(rowType)
+
     val writeBlocksPart = split.asInstanceOf[WriteBlocksRDDPartition]
     val start = writeBlocksPart.start
-    var i = 0
     writeBlocksPart.range.foreach { pi =>
-      val indexedRows = irm.rows.iterator(parentPartitions(pi), context)
+      val it = rdd2.rdd.iterator(parentParts(pi), context)
 
       if (pi == start) {
         var j = 0
         while (j < writeBlocksPart.skip) {
-          indexedRows.next()
+          it.next()
           j += 1
         }
       }
 
-      while (indexedRows.hasNext && i < nRowsInBlock) {
-        val indexedRow = indexedRows.next()
-        assert(indexedRow.index == firstRowInBlock + i,
-          s"IndexedRow index ${indexedRow.index} in partition $pi does not equal RDD index ${firstRowInBlock + i}")
-        
-        val data = indexedRow.vector.toArray // free on Spark DenseVector
-        Memory.memcpy(bytes, 0, data, 0, data.length)
-
-        var off = 0
+      var i = 0
+      while (it.hasNext && i < nRowsInBlock) {
+        val rv = it.next()
+        ur.set(rv)
+        ec.set(1, ur.get(1))
+        ec.set(2, ur.get(2))
+        val gs = ur.getAs[IndexedSeq[Any]](3)
+        var offset = 0
         var blockCol = 0
         while (blockCol < gp.nBlockCols) {
-          val n = gp.blockColNCols(blockCol) << 3
-          
-          dosArray(blockCol).write(bytes, off, n)
-          
-          off += blockSize << 3
-          blockCol+= 1
+          val n = gp.blockColNCols(blockCol)
+          var j = 0
+          while (j < n) {
+            ec.set(3, sampleIdsBc.value(j))
+            ec.set(4, sampleAnnotationsBc.value(j))
```

Comment:
You need `offset + j` in the sample annotations, too.  You might write this to be a bit safer: one counter for the offset in the block, another for the sample.

Reply:
Good catch. I've gone the safer route, and adjusted the test so it would have caught this.